contracts/hackatom/src/contract.rs

```diff
@@ -138,11 +143,7 @@ fn query_verifier<S: Storage, A: Api>(deps: &Extern<S, A>) -> Result<QueryRespon
         .context(NotFound { kind: "State" })?;
     let state: State = from_slice(&data)?;
     let addr = deps.api.human_address(&state.verifier)?;
-    // we just pass the address as raw bytes
-    // these will be base64 encoded into the json we return, and parsed on the way out.
-    // maybe we should wrap this in a struct then json encode it into a vec?
-    // other ideas?
-    Ok(Binary(addr.as_str().as_bytes().to_vec()))
+    Ok(Binary(to_vec(&VerifierResponse { verifier: addr })?))
```

Comment:
Ideally, we would enforce the JSON Serailizablilty in the `query` type.

Something like:

```rust
pub fn query<S: Storage, A: Api>(deps: &Extern<S, A>, msg: QueryMsg) -> Result<Box<dyn Serialize>> {
// ...
}
```

Note: `Box<dyn Serialize>` is needed to return an object, where we would accept `&dyn Serialize` as input. (THis allows runtime dynamic access to anything that implements the trait (as we don't define all possible types compile time) 

And then in the `_do_query` wraper, we can do

Reply:
I cannot get this to work. We need `Deserialize` as well. What I got is:

```rust
pub trait Serde<'a>: Serialize + Deserialize<'a> {}

pub type QueryResponse<'a> = Box<dyn Serde<'a>>;

pub type QueryResult<'a> = ApiResult<QueryResponse<'a>>;
```

which fails to compile with

```
error[E0038]: the trait `query::Serde` cannot be made into an object
   --> packages/std/src/exports.rs:120:1
    |
120 | / fn _do_query<T: DeserializeOwned + JsonSchema>(
121 | |     query_fn: &dyn Fn(&Extern<ExternalStorage, ExternalApi>, T) -> Result<QueryResponse, Error>,
122 | |     msg_ptr: *mut c_void,
123 | | ) -> Result<QueryResponse, Error> {
...   |
128 | |     query_fn(&deps, msg)
129 | | }
    | |_^ the trait `query::Serde` cannot be made into an object
    |
    = note: the trait cannot require that `Self : Sized`
    = note: method `serialize` has generic type parameters
```

and generating a schema for this also fails horribly. I don't feel capable doing this as long as there is nothing like [serde_json::value::RawValue](https://docs.serde.rs/serde_json/value/struct.RawValue.html) or [serde_json::Value](https://docs.serde.rs/serde_json/value/enum.Value.htmll) ready to be used.

Reply:
Fair enough. Let's not change it.
If I get motivated I will try later (and maybe fail again)